crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -956,11 +968,32 @@ pub fn is_const(expr: &Expression) -> bool {
     expr.mul_terms.is_empty() && expr.linear_combinations.is_empty()
 }
 
+pub fn mul_with_witness(evaluator: &mut Evaluator, a: &Expression, b: &Expression) -> Expression {
+    let a_arith;
+    let a2 = if !a.mul_terms.is_empty() {
+        let a_witness = evaluator.add_witness_to_cs();
+        a_arith = Expression::from(&a_witness);
+        evaluator.gates.push(Gate::Arithmetic(a - &a_arith));
+        &a_arith
+    } else {
+        a
+    };
+    let b_arith;
+    let b2 = if !b.mul_terms.is_empty() {
+        let b_witness = evaluator.add_witness_to_cs();
+        b_arith = Expression::from(&b_witness);
+        evaluator.gates.push(Gate::Arithmetic(b - &b_arith));
+        &b_arith
+    } else {
+        b
+    };
+    mul(a2, b2)
+}
 //a*b
 
 pub fn mul(a: &Expression, b: &Expression) -> Expression {
     if !(a.mul_terms.is_empty() && b.mul_terms.is_empty()) {
```

Comment:
Can you add a method to expression named `is_linear` instead of `mul_terms.is_empty()` ?

Reply:
Done